src/main/core/scheduler/pools/unbounded.rs

```diff
@@ -0,0 +1,474 @@
+use std::marker::PhantomData;
+use std::ops::Deref;
+use std::sync::atomic::{AtomicBool, Ordering};
+use std::sync::Arc;
+
+use atomic_refcell::AtomicRefCell;
+
+use crate::utility::synchronization::count_down_latch::{
+    build_count_down_latch, LatchCounter, LatchWaiter,
+};
+use crate::utility::synchronization::semaphore::LibcSemaphore;
+
+// If making substantial changes to this scheduler, you should uncomment each test at the end of
+// this file to make sure that they correctly cause a compilation error. This work pool unsafely
+// transmutes the task closure lifetime, and the commented tests are meant to make sure that the
+// work pool does not allow unsound code to compile. Due to lifetime sub-typing/variance, rust will
+// sometimes allow closures with shorter or longer lifetimes than we specify in the API, so the
+// tests check to make sure the closures are invariant over the lifetime and that the usage is
+// sound.
+
+/// A task that is run by the pool threads.
+pub trait TaskFn: Fn(usize) + Send + Sync {}
+impl<T> TaskFn for T where T: Fn(usize) + Send + Sync {}
+
+/// A thread pool that runs a task on many threads. A task will run once on each thread.
+pub struct UnboundedThreadPool {
+    /// Handles for joining threads when they've exited.
+    thread_handles: Vec<std::thread::JoinHandle<()>>,
+    /// State shared between all threads.
+    shared_state: Arc<SharedState>,
+    /// Semaphores used by work threads to wait for a new task.
+    task_start_semaphores: Vec<LibcSemaphore>,
+    /// The main thread uses this to wait for the threads to finish running the task.
+    task_end_waiter: LatchWaiter,
+}
+
+pub struct SharedState {
+    /// The task to run during the next round.
+    task: AtomicRefCell<Option<Box<dyn TaskFn>>>,
+    /// Has a thread panicked?
+    has_thread_panicked: AtomicBool,
+}
+
+impl UnboundedThreadPool {
+    pub fn new(num_threads: usize, thread_name: &str) -> Self {
+        let shared_state = Arc::new(SharedState {
+            task: AtomicRefCell::new(None),
+            has_thread_panicked: AtomicBool::new(false),
+        });
+
+        let (task_end_counter, task_end_waiter) = build_count_down_latch();
+
+        let mut thread_handles = Vec::new();
+        let mut task_start_semaphores = Vec::new();
+
+        for i in 0..num_threads {
+            let task_start_semaphore = LibcSemaphore::new(0);
```

Comment:
optional bikeshed, orthogonal to this PR - it's a little confusing that cloning below gives a reference to the same object. Maybe this type should be called something like `LibcSemaphoreArc`?

I also noticed the `SAFETY` comments in that type's definition - maybe it'd be worth using `Pin` to express in the type system that it can't be moved?

Reply:
I'm going to address this in a follow-up PR since I'll also want to change the thread-per-host scheduler at the same time.